src/RamOperation.cpp

```diff
@@ -30,187 +31,25 @@
-                if (arg != nullptr) {
-                    level = std::max(level, visit(arg));
-                }
-            }
-            return level;
-        }
-    };
-    return ValueLevelVisitor().visit(value);
-}
-
-/** Get level of condition (which for-loop of a query) */
-// TODO (#541): Remove or replace by RamConditionLevelAnalysis
-size_t getLevel(const RamCondition* condition) {
-    // visitor
-    class ConditionLevelVisitor : public RamVisitor<size_t> {
-    public:
-        // conjunction
-        size_t visitConjunction(const RamConjunction& conj) override {
-            return std::max(visit(conj.getLHS()), visit(conj.getRHS()));
-        }
-
-        // negation
-        size_t visitNegation(const RamNegation& neg) override {
-            return visit(neg.getOperand());
-        }
-
-        // binary constraint
-        size_t visitConstraint(const RamConstraint& binRel) override {
-            return std::max(getLevel(binRel.getLHS()), getLevel(binRel.getRHS()));
-        }
-
-        // not exists check
-        size_t visitExistenceCheck(const RamExistenceCheck& exists) override {
-            size_t level = 0;
-            for (const auto& cur : exists.getValues()) {
-                if (cur != nullptr) {
-                    level = std::max(level, getLevel(cur));
-                }
-            }
-            return level;
-        }
-
-        // not exists check for a provenance existence check
-        size_t visitProvenanceExistenceCheck(const RamProvenanceExistenceCheck& provExists) override {
-            size_t level = 0;
-            for (const auto& cur : provExists.getValues()) {
-                if (cur != nullptr) {
-                    level = std::max(level, getLevel(cur));
-                }
-            }
-            return level;
-        }
-
-        // emptiness check
-        size_t visitEmptinessCheck(const RamEmptinessCheck& emptiness) override {
-            return 0;  // can be in the top level
-        }
-    };
-    return ConditionLevelVisitor().visit(condition);
-}
-
 /** get indexable element */
-std::unique_ptr<RamExpression> RamAggregate::getIndexElement(RamCondition* c, size_t& element, size_t level) {
-    if (auto* binRelOp = dynamic_cast<RamConstraint*>(c)) {
+std::unique_ptr<RamExpression> RamAggregate::getIndexElement(
+        RamCondition* condition, size_t& element, size_t level) {
+    if (auto* binRelOp = dynamic_cast<RamConstraint*>(condition)) {
         if (binRelOp->getOperator() == BinaryConstraintOp::EQ) {
             if (auto* lhs = dynamic_cast<RamElementAccess*>(binRelOp->getLHS())) {
                 RamExpression* rhs = binRelOp->getRHS();
-                if (lhs->getIdentifier() == level && (isConstant(rhs) || getLevel(rhs) < level)) {
+                if (lhs->getIdentifier() == level &&
+                        (RamConstValueAnalysis().isConstant(rhs) ||
+                                RamExpressionLevelAnalysis().getLevel(rhs) < level)) {
```

Comment:
I think all construction of RAM analyses are meant to go through RamTranslationUnit.getAnalysis(). This way looks good until that happens, though - can remove a bunch of duplicate code.

Reply:
getAnalysis does a nice caching thing with the analyses, but these in particular have no state so there's no benefit from the caching.

It does mean there are used differently, though. If we require that all analyses are retrieved from the RamTranslationUnit, then RamAggregate and AstTranslator should be rewritten so that RamAggregate::addCondition() just adds a condition, rather than processing it, then adding the result.